### DISCUSSION TITLE
Is there a better constraint modeling method under Logic programming？

### DISCUSSION THREAD

**Discussion by lzmssyld** [yalmip/YALMIP#1434] [General discussion] [answered]
Dear sir！
1. Problem description
          I modeled a scheduling problem as a mixed integer programming model. I used MATLAB+YALMIP+GUROBI to solve the model.
  In the model, I need to construct the following logic as Constraints 1 and 2：
  If  a(1)<=b(5)
      a(2)<=b(5)
  else
      a(2)<=b(6)
  end 
         As the dimensionality（number_mission） of variables increases, model constraints will dramatically increase, resulting in longer solution times.
         Can constraint 1 and constraint 2 be modeled in a more efficient way? 
2. Model code
```
% Variable Definitions
x_d=intvar(number_mission,num_cp_1,'full');% 
x_a=intvar(number_mission,num_cp_2,'full');%
% Constraint modeling
% Constraint 1
for i=1:number_mission
    for j=1:number_mission
       if i ~= j 
     F=[F,implies(x_a(i,1)<=x_d(j,5),x_a(i,2)<=x_d(j,5))];
       end 
    end
end
% Constraint 2
for i=1:number_mission
    for j=1:number_mission
       if i ~= j 
     F=[F,implies(x_d(i,5)<=x_a(j，1)-0.01,x_a(j,2)<=x_d(i,6))];
       end 
    end
end

```

**Comment by johanlofberg** [comment, marked as answer]
As explained in the logic programming pages, this is not a recommended way to model that disjunction, as it could be that neither of the conditions are triggered due to numerical tolerances in the solver.

The way to do it robustly is to explicitly list the disjoint cases and trigger with binary variables caseq/case2
`
  F=[F,implies(case1(i), [x_a(i,1)<=x_d(j,5),x_a(i,2)<=x_d(j,5)]) + implies(case2(i),[x_a(i,1)>=x_d(j,5)]) +[case1(i)+case2(i) == 1]`

You're not going to get it faster though, generically, as you have a large combinatorial problem

**Reply by lzmssyld** [reply]
I appreciate greatly your assistance.  Thank you for your time and attention！

**Comment by lzmssyld** [comment]
I have now resolved PART of  my problem by restructuring the problem model to avoid using “implies” in PART of constraints，this greatly reduces the number of constraints.
